File: src/main/java/org/team401/taxis/diffdrive/control/DrivetrainPathManager.java
package org.team401.taxis.diffdrive.control;

import org.team401.taxis.geometry.Pose2d;
import org.team401.taxis.geometry.Pose2dWithCurvature;
import org.team401.taxis.geometry.Rotation2d;
import org.team401.taxis.physics.DifferentialDrive;
import org.team401.taxis.trajectory.*;
import org.team401.taxis.trajectory.timing.DifferentialDriveDynamicsConstraint;
import org.team401.taxis.trajectory.timing.TimedState;
import org.team401.taxis.trajectory.timing.TimingConstraint;
import org.team401.taxis.trajectory.timing.TimingUtil;
import org.team401.taxis.util.CSVWritable;
import org.team401.taxis.util.Units;
import org.team401.taxis.util.Util;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Cameron Earle
 * @version 12/3/2018
 *
 */
public class DrivetrainPathManager implements CSVWritable {
    public DrivetrainPathManager(FullStateDiffDriveModel fullStateModel,
                                 PathController controller,
                                 double maxDx,
                                 double maxDy,
                                 double maxDTheta) {
        this.fullStateModel = fullStateModel;
        this.controller = controller;
        kMaxDx = maxDx;
        kMaxDy = maxDy;
        kMaxDTheta = maxDTheta;
    }

    private final FullStateDiffDriveModel fullStateModel;
    private PathController controller;

    private double kMaxDx;
    private double kMaxDy;
    private double kMaxDTheta;

    TrajectoryIterator<TimedState<Pose2dWithCurvature>> mCurrentTrajectory;
    boolean mIsReversed = false;
    double mLastTime = Double.POSITIVE_INFINITY;
    TimedState<Pose2dWithCurvature> mSetpoint = new TimedState<>(Pose2dWithCurvature.identity());
    Pose2d mError = Pose2d.identity();
    Output mOutput = new Output();

    double mDt = 0.0;

    public Trajectory<TimedState<Pose2dWithCurvature>> generateTrajectory(
            boolean reversed,
            final List<Pose2d> waypoints,
            final List<TimingConstraint<Pose2dWithCurvature>> constraints,
            double max_vel,  // inches/s
            double max_accel,  // inches/s^2
            double max_voltage) {
        return generateTrajectory(reversed, waypoints, constraints, 0.0, 0.0, max_vel, max_accel, max_voltage);
    }

    public Trajectory<TimedState<Pose2dWithCurvature>> generateTrajectory(
            boolean reversed,
            final List<Pose2d> waypoints,
            final List<TimingConstraint<Pose2dWithCurvature>> constraints,
            double start_vel,
            double end_vel,
            double max_vel,  // inches/s
            double max_accel,  // inches/s^2
            double max_voltage) {

        List<Pose2d> waypoints_maybe_flipped = waypoints;
        final Pose2d flip = Pose2d.fromRotation(new Rotation2d(-1, 0, false));
        // TODO re-architect the spline generator to support reverse.
        if (reversed) {
            waypoints_maybe_flipped = new ArrayList<>(waypoints.size());
            for (int i = 0; i < waypoints.size(); ++i) {
                waypoints_maybe_flipped.add(waypoints.get(i).transformBy(flip));
            }
        }

        // Create a trajectory from splines.
        Trajectory<Pose2dWithCurvature> trajectory = TrajectoryUtil.trajectoryFromSplineWaypoints(
                waypoints_maybe_flipped, kMaxDx, kMaxDy, kMaxDTheta);

        if (reversed) {
            List<Pose2dWithCurvature> flipped = new ArrayList<>(trajectory.length());
            for (int i = 0; i < trajectory.length(); ++i) {
                flipped.add(new Pose2dWithCurvature(trajectory.getState(i).getPose().transformBy(flip), -trajectory
                        .getState(i).getCurvature(), trajectory.getState(i).getDCurvatureDs()));
            }
            trajectory = new Trajectory<>(flipped);
        }
        // Create the constraint that the robot must be able to traverse the trajectory without ever applying more
        // than the specified voltage.
        final DifferentialDriveDynamicsConstraint<Pose2dWithCurvature> drive_constraints = new
                DifferentialDriveDynamicsConstraint<>(fullStateModel.getDrivetrainDynamicsModel(), max_voltage);
        List<TimingConstraint<Pose2dWithCurvature>> all_constraints = new ArrayList<>();
        all_constraints.add(drive_constraints);
        if (constraints != null) {
            all_constraints.addAll(constraints);
        }
        // Generate the timed trajectory.
        Trajectory<TimedState<Pose2dWithCurvature>> timed_trajectory = TimingUtil.timeParameterizeTrajectory
                (reversed, new
                        DistanceView<>(trajectory), kMaxDx, all_constraints, start_vel, end_vel, max_vel, max_accel);
        return timed_trajectory;
    }

    public void setTrajectory(final TrajectoryIterator<TimedState<Pose2dWithCurvature>> trajectory) {
        mCurrentTrajectory = trajectory;
        mSetpoint = trajectory.getState();
        for (int i = 0; i < trajectory.trajectory().length(); ++i) {
            if (trajectory.trajectory().getState(i).velocity() > Util.kEpsilon) {
                mIsReversed = false;
                break;
            } else if (trajectory.trajectory().getState(i).velocity() < -Util.kEpsilon) {
                mIsReversed = true;
                break;
            }
        }
    }

    @Override
    public String toCSV() {
        DecimalFormat fmt = new DecimalFormat("#0.000");
        return fmt.format(mOutput.left_velocity) + "," + fmt.format(mOutput.right_velocity) + "," + fmt.format
                (mOutput.left_feedforward_voltage) + "," + fmt.format(mOutput.right_feedforward_voltage) + "," +
                mSetpoint.toCSV();
    }

    public Output update(double timestamp, Pose2d current_state) {
        if (mCurrentTrajectory == null) return new Output();

        if (mCurrentTrajectory.getProgress() == 0.0 && !Double.isFinite(mLastTime)) {
            mLastTime = timestamp;
        }

        mDt = timestamp - mLastTime;
        mLastTime = timestamp;
        TrajectorySamplePoint<TimedState<Pose2dWithCurvature>> sample_point = mCurrentTrajectory.advance(mDt);
        mSetpoint = sample_point.state();

        if (!mCurrentTrajectory.isDone()) {
            // Generate feedforward voltages.
            final double velocity_m = Units.inches_to_meters(mSetpoint.velocity());
            final double curvature_m = Units.meters_to_inches(mSetpoint.state().getCurvature());
            final double dcurvature_ds_m = Units.meters_to_inches(Units.meters_to_inches(mSetpoint.state()
                    .getDCurvatureDs()));
            final double acceleration_m = Units.inches_to_meters(mSetpoint.acceleration());
            final DifferentialDrive.DriveDynamics dynamics = fullStateModel.getDrivetrainDynamicsModel().solveInverseDynamics(
                    new DifferentialDrive.ChassisState(velocity_m, velocity_m * curvature_m),
                    new DifferentialDrive.ChassisState(acceleration_m,
                            acceleration_m * curvature_m + velocity_m * velocity_m * dcurvature_ds_m));
            mError = current_state.inverse().transformBy(mSetpoint.state().getPose());

            mOutput = controller.update(dynamics, current_state, mError, fullStateModel.getDrivetrainDynamicsModel(), mDt);
        } else {
            // TODO Possibly switch to a pose stabilizing controller?
            mOutput = new Output();
        }
        return mOutput;
    }

    public boolean isDone() {
        return mCurrentTrajectory != null && mCurrentTrajectory.isDone();
    }

    public Pose2d getError() {
        return mError;
    }

    public TimedState<Pose2dWithCurvature> getSetpoint() {
        return mSetpoint;
    }

    public void reset() {
        mError = Pose2d.identity();
        mOutput = new Output();
        mLastTime = Double.POSITIVE_INFINITY;
    }

    public synchronized PathController getController() {
        return controller;
    }

    public synchronized void setController(PathController controller) {
        this.controller = controller;
    }

    public static class Output {
        public Output() {
        }

        public Output(double left_velocity, double right_velocity, double left_accel, double right_accel,
                      double left_feedforward_voltage, double
                              right_feedforward_voltage) {
            this.left_velocity = left_velocity;
            this.right_velocity = right_velocity;
            this.left_accel = left_accel;
            this.right_accel = right_accel;
            this.left_feedforward_voltage = left_feedforward_voltage;
            this.right_feedforward_voltage = right_feedforward_voltage;
        }

        public double left_velocity;  // rad/s
        public double right_velocity;  // rad/s

        public double left_accel;  // rad/s^2
        public double right_accel;  // rad/s^2

        public double left_feedforward_voltage;
        public double right_feedforward_voltage;

        public void flip() {
            double tmp_left_velocity = left_velocity;
            left_velocity = -right_velocity;
            right_velocity = -tmp_left_velocity;

            double tmp_left_accel = left_accel;
            left_accel = -right_accel;
            right_accel = -tmp_left_accel;

            double tmp_left_feedforward = left_feedforward_voltage;
            left_feedforward_voltage = -right_feedforward_voltage;
            right_feedforward_voltage = -tmp_left_feedforward;
        }
    }
}

File: src/main/kotlin/org/team401/taxis/diffdrive/control/DrivetrainPathManager.kt
package org.team401.taxis.diffdrive.control

import org.team401.taxis.diffdrive.Path
import org.team401.taxis.geometry.Pose2d
import org.team401.taxis.geometry.Pose2dWithCurvature
import org.team401.taxis.geometry.Rotation2d
import org.team401.taxis.physics.DifferentialDrive
import org.team401.taxis.template.PathFollowingTemplate
import org.team401.taxis.trajectory.*
import org.team401.taxis.trajectory.timing.DifferentialDriveDynamicsConstraint
import org.team401.taxis.trajectory.timing.TimedState
import org.team401.taxis.trajectory.timing.TimingConstraint
import org.team401.taxis.trajectory.timing.TimingUtil
import org.team401.taxis.util.Units
import org.team401.taxis.util.Util

/**
 * @author Cameron Earle
 * @version 9/10/2018
 *
 * Manages path following for a drivetrain.  This class does not interact with any hardware,
 * and instead must be provided with the inputs by calling appropriate functions.
 *
 * @param dynamicsModel The dynamics model of the drivetrain
 * @param pathFollowingConfig The path following data
 * @param controller The drive controller to use
 */
class DrivetrainPathManager(private val dynamicsModel: DifferentialDrive,
                            private val pathFollowingConfig: PathFollowingTemplate,
                            private val controller: PathController) {

    private val trajectories = hashMapOf<String, Trajectory<TimedState<Pose2dWithCurvature>>>()

    //Some process variables
    private var currentTrajectory: TrajectoryIterator<TimedState<Pose2dWithCurvature>>? = null
    private var reversed = false
    private var lastTime = Double.POSITIVE_INFINITY
    var setpoint = TimedState<Pose2dWithCurvature>(Pose2dWithCurvature.identity())
        @Synchronized get
        private set
    var error = Pose2d.identity()
        @Synchronized get
        private set
    private var output = Output()
    private var dt = 0.0

    @Synchronized fun reset() {
        error = Pose2d.identity()
        output = Output()
        lastTime = Double.POSITIVE_INFINITY
        controller.reset()
    }

    /**
     * Generates a trajectory from a path
     */
    private fun generateTrajectory(path: Path): Trajectory<TimedState<Pose2dWithCurvature>> {
        val flip = Pose2d.fromRotation(Rotation2d(-1.0, 0.0, false))
        val waypoints = if (reversed) {
            path.waypoints.map { it.transformBy(flip) }
        } else {
            path.waypoints
        }

        var trajectory = TrajectoryUtil.trajectoryFromSplineWaypoints(
                waypoints,
                pathFollowingConfig.maxErrorX,
                pathFollowingConfig.maxErrorY,
                pathFollowingConfig.maxErrorTheta
        )

        if (reversed) {
            val flipped = ArrayList<Pose2dWithCurvature>(trajectory.length())
            for (i in 0 until trajectory.length()) {
                flipped.add(
                        Pose2dWithCurvature(
                                trajectory.getState(i).pose.transformBy(flip),
                                -trajectory.getState(i).curvature,
                                trajectory.getState(i).dCurvatureDs
                        )
                )
            }
            trajectory = Trajectory(flipped)
        }

        val driveConstraints = DifferentialDriveDynamicsConstraint<Pose2dWithCurvature>(dynamicsModel, path.maxVoltage)
        val allConstraints = arrayListOf<TimingConstraint<Pose2dWithCurvature>>()
        allConstraints.add(driveConstraints)
        allConstraints.addAll(path.constraints)

        return TimingUtil.timeParameterizeTrajectory(
                reversed,
                DistanceView(trajectory),
                pathFollowingConfig.maxErrorX,
                allConstraints,
                path.startVelocity,
                path.endVelocity,
                path.maxVelocity,
                path.maxAcceleration
        )
    }

    /**
     * Sets the current trajectory to follow, given an iterator
     */
    private fun setTrajectory(trajectory: TrajectoryIterator<TimedState<Pose2dWithCurvature>>) {
        currentTrajectory = trajectory
        setpoint = trajectory.state
        for (i in 0 until trajectory.trajectory().length()) {
            if (trajectory.trajectory().getState(i).velocity() > Util.kEpsilon) {
                reversed = false
                break
            } else if (trajectory.trajectory().getState(i).velocity() < -Util.kEpsilon) {
                reversed = true
                break
            }
        }
    }

    /**
     * Adds a path to the path manager
     *
     * @param name The name of the path.  Will be used to recall the path to follow it
     * @param path The path
     */
    @Synchronized fun addPath(name: String, path: Path) {
        trajectories[name] = generateTrajectory(path)
    }

    /**
     * Sets the path by the given name as the current path.
     *
     * @param name The name of the path to follow
     */
    @Synchronized fun setPath(name: String) {
        if (trajectories.contains(name)) {
            val iterator = TrajectoryIterator(TimedView(trajectories[name]))
            setTrajectory(iterator)
        } else {
            throw NoSuchElementException("Could not find path '$name'")
        }
    }

    /**
     * Updates the path follower with new state information.
     * This information is passed to the controller
     */
    @Synchronized fun update(timestamp: Double, currentState: Pose2d): Output {
        if (currentTrajectory == null) return Output()
        val currentTrajectory = currentTrajectory!! //Bypass null checks

        if (currentTrajectory.progress == 0.0 && !java.lang.Double.isFinite(lastTime)) {
            lastTime = timestamp
        }

        dt = timestamp - lastTime
        lastTime = timestamp
        val samplePoint = currentTrajectory.advance(dt)
        setpoint = samplePoint.state()

        if (!currentTrajectory.isDone) {
            //Generate feedforward voltages
            val velocityM = Units.inches_to_meters(setpoint.velocity())
            val curvatureM = Units.meters_to_inches(setpoint.state().curvature)
            val dCurvatureDsM = Units.meters_to_inches(Units.meters_to_inches(setpoint.state().dCurvatureDs))
            val accelerationM = Units.inches_to_meters(setpoint.acceleration())
            val dynamics = dynamicsModel.solveInverseDynamics(
                    DifferentialDrive.ChassisState(velocityM, velocityM * curvatureM),
                    DifferentialDrive.ChassisState(accelerationM,
                            accelerationM * curvatureM + velocityM * velocityM * dCurvatureDsM)
            )
            error = currentState.inverse().transformBy(setpoint.state().pose)
            output = controller.update(
                    dt,
                    dynamics,
                    currentState,
                    dynamicsModel,
                    currentTrajectory,
                    setpoint,
                    error,
                    reversed
            )
        } else {
            output = Output()
        }

        return output
    }

    @Synchronized fun isDone() = currentTrajectory != null && currentTrajectory!!.isDone
}